README.md
# OpenRosa FormSubmissionAPI middleware
[![travis][travis-image]][travis-url]
[![js-standard-style][js-standard-image]][js-standard-url]
[![npm][npm-image]][npm-url]
[![downloads][downloads-image]][downloads-url]

[travis-image]: https://travis-ci.org/digidem/openrosa-form-submission-middleware.svg?branch=master
[travis-url]: https://travis-ci.org/digidem/openrosa-form-submission-middleware
[js-standard-image]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg
[js-standard-url]: http://standardjs.com/
[npm-image]: https://img.shields.io/npm/v/openrosa-form-submission-middleware.svg?style=flat
[npm-url]: https://npmjs.org/package/openrosa-form-submission-middleware
[downloads-image]: https://img.shields.io/npm/dm/openrosa-form-submission-middleware.svg?style=flat
[downloads-url]: https://npmjs.org/package/openrosa-form-submission-middleware

This is based on
[connect-multiparty](https://github.com/andrewrk/connect-multiparty) by
[Andrew Kelley](https://github.com/andrewrk/).

It is [express](http://expressjs.com/) middleware for [multiparty](https://github.com/andrewrk/node-multiparty/) to process OpenRosa form submissions from [ODK Collect](https://opendatakit.org/use/collect/) following the [OpenRosa FormSubmissionAPI spec](https://bitbucket.org/javarosa/javarosa/wiki/FormSubmissionAPI).

The xml form submission is returned as req.body and any attached files are returned as req.files.

Incoming files are stored on disk in the `tmp` folder and must be cleanup up afterwards with something like:

```js
req.files.forEach(function(file) {
    fs.unlink(file.path, function() {});
});
```

## Usage

```js
var openrosa = require('openrosa-form-submission-middleware');
var openrosaMiddleware = openrosa();

app.use('/submission', openrosaMiddleware);

app.post('/submission', function(req, res) {
  console.log(req.body, req.files);
  // don't forget to delete all req.files when done
});
```

## API

### openrosa(options)

Returns express middleware for receiving and processing OpenRosa form submissions.

* `options.maxContentLength` sets the maximum content length of form submissions (defaults to 10Mb)

* `options.secure` will redirect OpenRosa clients like ODK Collect to use https to send a submission (it does this by responding to the initial HEAD request from the client with a 204 with the Location headers set with the https protocol set)

Other options are passed directly on to multiparty.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [1.1.0]
### Fixed
- Set location header on response to HEAD request fixes #1

### Added
- Add option to redirect to https for secure submission

## [1.0.4]
### Changed
- Switch to [JS Standard style](https://github.com/feross/standard)
- Add debugging messages

## [1.0.3]
### Changed
- Minor fixes to code style and error reporting

## [1.0.2]
### Fixed
- Correctly respond to HEAD requests

### Changed
- Switch to Express for testing
- Add debug messages

## [1.0.1]
### Fixed
- Remove form from disk once processed

### 1.0.0

Initial release

[Unreleased]: https://github.com/digidem/openrosa-form-submission-middleware/compare/v1.1.0...HEAD
[1.1.0]: https://github.com/digidem/openrosa-form-submission-middleware/compare/v1.0.4...v1.1.0
[1.0.4]: https://github.com/digidem/openrosa-form-submission-middleware/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/digidem/openrosa-form-submission-middleware/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/digidem/openrosa-form-submission-middleware/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/digidem/openrosa-form-submission-middleware/compare/v1.0.0...v1.0.1


test/openrosa-form-submission.js
/*global describe it*/
process.env.NODE_ENV = 'test'

var express = require('express')
var formSubmission = require('..')
var request = require('supertest')
var fs = require('fs')
var path = require('path')

require('should')

var app = express()

app.use(formSubmission())

app.use(function (req, res) {
  res.end(JSON.stringify(req.body))
})

describe('formSubmissionMiddleware()', function () {
  it('should ignore GET', function (done) {
    request(app)
      .get('/')
      .set('X-OpenRosa-Version', '1.0')
      .field('user', 'Tobi')
      .expect(200, '', done)
  })

  it('should respond to HEAD with 204 and X-OpenRosa-Version header', function (done) {
    request(app)
      .head('/')
      .set('X-OpenRosa-Version', '1.0')
      .field('user', 'Tobi')
      .expect('X-OpenRosa-Version', '1.0')
      .expect(204, '', done)
  })

  it('should include Location header pointing to same url in response to HEAD request', function (done) {
    var app = express()
    app.use(function (req, res, next) {
      req.connection.encrypted = false
      req.headers.host = 'example.com'
      next()
    })
    app.use(formSubmission())

    request(app)
      .head('/')
      .set('X-OpenRosa-Version', '1.0')
      .expect('X-OpenRosa-Version', '1.0')
      .expect('Location', 'http://example.com/')
      .expect(204, '', done)
  })

  it('encrypted connections should respond with Location header set to https too', function (done) {
    var app = express()
    app.use(function (req, res, next) {
      req.connection.encrypted = true
      req.headers.host = 'example.com'
      next()
    })
    app.use(formSubmission())

    request(app)
      .head('/')
      .set('X-OpenRosa-Version', '1.0')
      .expect('X-OpenRosa-Version', '1.0')
      .expect('Location', 'https://example.com/')
      .expect(204, '', done)
  })

  it('Setting options.secure should cause head request to redirect to https', function (done) {
    var app = express()
    app.use(function (req, res, next) {
      req.connection.encrypted = false
      req.headers.host = 'example.com'
      next()
    })
    app.use(formSubmission({secure: true}))

    request(app)
      .head('/')
      .set('X-OpenRosa-Version', '1.0')
      .expect('X-OpenRosa-Version', '1.0')
      .expect('Location', 'https://example.com/')
      .expect(204, '', done)
  })

  describe('with multipart/form-data', function () {
    var xmlFixture = fs.readFileSync(path.join(__dirname, 'fixtures', 'xml_submission_file.xml'))

    it('should return xml_submission_file as the form body', function (done) {
      var app = express()

      app.use(formSubmission())

      app.use(function (req, res) {
        req.body.should.eql(xmlFixture.toString())
        req.files.length.should.eql(0)
        res.end('{}')
      })

      request(app)
        .post('/')
        .set('X-OpenRosa-Version', '1.0')
        .attach('xml_submission_file', xmlFixture, { filename: 'foo.xml' })
        .expect(200, '{}', done)
    })

    it('should handle multiple file attachments', function (done) {
      var app = express()

      app.use(formSubmission())

      app.use(function (req, res) {
        req.body.should.eql(xmlFixture.toString())
        req.files[0].path.should.endWith('.jpg')
        req.files.length.should.eql(2)
        res.end(req.files[1].originalFilename)
      })

      request(app)
        .post('/')
        .set('X-OpenRosa-Version', '1.0')
        .attach('xml_submission_file', xmlFixture, { filename: 'foo.xml' })
        .attach('photo', path.join(__dirname, 'fixtures', 'image.jpg'))
        .attach('photo', path.join(__dirname, 'fixtures', 'image2.jpg'))
        .expect(200, 'image2.jpg', done)
    })

    it('should support multiple files of the same name', function (done) {
      var app = express()

      app.use(formSubmission())

      app.use(function (req, res) {
        req.files.should.have.length(2)
        req.files[0].constructor.name.should.equal('Object')
        req.files[1].constructor.name.should.equal('Object')
        res.end()
      })

      request(app)
        .post('/')
        .set('X-OpenRosa-Version', '1.0')
        .attach('xml_submission_file', xmlFixture, { filename: 'foo.xml' })
        .attach('text', new Buffer('some text here'), { filename: 'foo.txt' })
        .attach('text', new Buffer('some more text stuff'), { filename: 'bar.txt' })
        .expect(200, done)
    })

    it('should support nested files', function (done) {
      var app = express()

      app.use(formSubmission())

      app.use(function (req, res) {
        req.files.should.have.length(2)
        req.files[0].originalFilename.should.equal('foo.txt')
        req.files[1].originalFilename.should.equal('bar.txt')
        res.end()
      })

      request(app)
        .post('/')
        .set('X-OpenRosa-Version', '1.0')
        .attach('xml_submission_file', xmlFixture, { filename: 'foo.xml' })
        .attach('docs[foo]', new Buffer('some text here'), { filename: 'foo.txt' })
        .attach('docs[bar]', new Buffer('some more text stuff'), { filename: 'bar.txt' })
        .expect(200, done)
    })

    it('should next(err) on multipart failure', function (done) {
      var app = express()

      app.use(formSubmission())

      app.use(function (req, res) {
        res.end('whoop')
      })

      app.use(function (err, req, res, next) {
        err.message.should.equal('Expected alphabetic character, received 61')
        res.statusCode = err.status
        res.end('bad request')
      })

      var test = request(app).post('/').set('X-OpenRosa-Version', '1.0')
      test.set('Content-Type', 'multipart/form-data; boundary=foo')
      test.write('--foo\r\n')
      test.write('Content-filename="foo.txt"\r\n')
      test.write('\r\n')
      test.write('some text here')
      test.write('Content-Disposition: form-data; name="text"; filename="bar.txt"\r\n')
      test.write('\r\n')
      test.write('some more text stuff')
      test.write('\r\n--foo--')
      test.expect(400, 'bad request', done)
    })

    it('should not hang request on failure', function (done) {
      var app = express()
      var buf = new Buffer(1024 * 10)

      app.use(formSubmission())

      app.use(function (req, res) {
        res.end('whoop')
      })

      app.use(function (err, req, res, next) {
        err.message.should.equal('Expected alphabetic character, received 61')
        res.statusCode = err.status
        res.end('bad request')
      })

      buf.fill('.')

      var test = request(app).post('/').set('X-OpenRosa-Version', '1.0')
      test.set('Content-Type', 'multipart/form-data; boundary=foo')
      test.write('--foo\r\n')
      test.write('Content-filename="foo.txt"\r\n')
      test.write('\r\n')
      test.write('some text here')
      test.write('Content-Disposition: form-data; name="text"; filename="bar.txt"\r\n')
      test.write('\r\n')
      test.write('some more text stuff')
      test.write('\r\n--foo--')
      test.write(buf)
      test.write(buf)
      test.write(buf)
      test.expect(400, 'bad request', done)
    })

    it('should default req.files to []', function (done) {
      var app = express()

      app.use(formSubmission())

      app.use(function (req, res) {
        res.end(JSON.stringify(req.files))
      })

      request(app)
        .post('/')
        .set('X-OpenRosa-Version', '1.0')
        .expect(200, '[]', done)
    })

    it('should return 400 on maxFilesSize exceeded', function (done) {
      var app = express()

      var exp = 9
      app.use(formSubmission({ maxFilesSize: Math.pow(2, exp) }))

      app.use(function (req, res) {
        res.end(JSON.stringify(req.files))
      })

      var str = 'x'
      for (var i = 0; i < exp + 1; i += 1) {
        str += str
      }

      request(app)
        .post('/')
        .set('X-OpenRosa-Version', '1.0')
        .field('user[name]', 'Tobi')
        .attach('text', new Buffer(str), { filename: 'foo.txt' })
        .expect(400, done)
    })

    it('should return next(err) if no xml_submission_file is included', function (done) {
      var app = express()

      app.use(formSubmission())

      app.use(function (req, res) {
        res.end('whoop')
      })

      app.use(function (err, req, res, next) {
        err.message.should.equal('No xml submission file included in request')
        res.statusCode = err.status
        res.end('bad request')
      })

      request(app)
        .post('/')
        .set('X-OpenRosa-Version', '1.0')
        .attach('photo', path.join(__dirname, 'fixtures', 'image.jpg'))
        .expect(400, 'bad request', done)
    })

    it('should reject requests without X-OpenRosa-Version header', function (done) {
      var app = express()

      app.use(formSubmission())

      app.use(function (req, res) {
        res.end('whoop')
      })

      request(app)
        .post('/')
        .attach('xml_submission_file', xmlFixture, { filename: 'foo.xml' })
        .expect(400, done)
    })

    it('should include "X-OpenRosa-Accept-Content-Length" header in response', function (done) {
      var app = express()

      app.use(formSubmission())

      app.use(function (req, res) {
        res.end('whoop')
      })

      request(app)
        .post('/')
        .set('X-OpenRosa-Version', '1.0')
        .expect('X-OpenRosa-Accept-Content-Length', '10485760')
        .attach('xml_submission_file', xmlFixture, { filename: 'foo.xml' })
        .expect(200, done)
    })

  })
})


index.js
/*!
 * Connect - multipart
 * Copyright(c) 2010 Sencha Inc.
 * Copyright(c) 2011 TJ Holowaychuk
 * Copyright(c) 2013 Andrew Kelley
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var multiparty = require('multiparty')
var onFinished = require('on-finished')
var typeis = require('type-is')
var fs = require('fs')
var debug = require('debug')('openrosa:submission')
var openrosaRequest = require('openrosa-request-middleware')
var openrosaRequestMiddleware = openrosaRequest()

/**
 * Multipart:
 *
 * Parse multipart/form-data request bodies,
 * providing the parsed object as `req.body`
 * and `req.files`.
 *
 * Configuration:
 *
 *  The options passed are merged with [multiparty](https://github.com/andrewrk/node-multiparty)'s
 *  `Form` object, allowing you to configure the upload directory,
 *  size limits, etc. For example if you wish to change the upload dir do the following.
 *
 *     app.use(connect.multipart({ uploadDir: path }))
 *
 * @param {Object} options
 * @return {Function}
 * @api public
 */

module.exports = function (options) {
  options = options || {}
  options.maxContentLength = options.maxContentLength || 10485760

  function multipart (req, res, next) {
    if (req._body) return next()
    req.files = []
    req.body = undefined

    if (req.method === 'HEAD') {
      // Response to HEAD requests with status 204
      return res.sendStatus(204)
    } else if (req.method !== 'POST') {
      // Ignore all other request methods other than post
      return next()
    }

    // check Content-Type
    if (!typeis(req, 'multipart/form-data')) return next()

    debug('processing form submission')

    // flag as parsed
    req._body = true

    // parse
    var form = new multiparty.Form(options)
    var done = false
    var processingXml = false
    var wasError = false

    form.on('file', function (name, val) {
      val.name = val.originalFilename
      val.type = val.headers['content-type'] || null

      if (name === 'xml_submission_file') {
        debug('received xml submission file (%s bytes)', val.size)
        processingXml = true
        fs.readFile(val.path, function (err, data) {
          if (err) onError(err)
          req.body = data.toString()
          fs.unlink(val.path, function (err) {
            if (err) console.error('Error deleting file %s', val.path)
          })
          debug('processed xml submission')
          processingXml = false
          if (done && !wasError) next()
        })

      } else {
        debug('received attachment %s (%s bytes)', val.name, val.size)
        req.files.push(val)
      }
    })

    form.on('error', function (err) {
      if (done) return
      onError(err)
    })

    form.on('close', function () {
      if (done) return

      done = true

      // only continue if we have already processed the xml submission file
      // and attached it to the req.body
      if (req.body) {
        next()
      } else if (!processingXml) {
        onError(new Error('No xml submission file included in request'))
      }
    })

    function onError (err) {
      debug('form processing error:', err.message)
      done = wasError = true

      err.status = 400

      if (!req.readable) return next(err)

      req.resume()
      onFinished(req, function () {
        next(err)
      })
    }

    form.parse(req)
  }

  return function (req, res, next) {
    // Set correct OpenRosa headers
    // see https://bitbucket.org/javarosa/javarosa/wiki/OpenRosaRequest
    // and https://bitbucket.org/javarosa/javarosa/wiki/FormSubmissionAPI
    openrosaRequestMiddleware(req, res, function (err) {
      if (err) next(err)
      res.setHeader('X-OpenRosa-Accept-Content-Length', options.maxContentLength)
      multipart(req, res, next)
    })
  }
}
